smarts/core/sensors.py

```diff
@@ -419,12 +422,12 @@ def _vehicle_is_off_road(cls, sim, vehicle):
 
     @classmethod
     def _vehicle_is_on_shoulder(cls, sim, vehicle):
-        return any(
-            [
-                not sim.scenario.road_network.point_is_within_road(corner_coordinate)
-                for corner_coordinate in vehicle.bounding_box
-            ]
-        )
+        # XXX: this isn't technically right as this would also return True
+        #      for vehicles that are completely off road.
+        for corner_coordinate in vehicle.bounding_box:
+            if not sim.scenario.road_network.point_is_within_road(corner_coordinate):
+                return True
+        return False
```

Comment:
The above is an unrelated (and very minor) optimization that I saw while looking at flamegraphs for the waypoints cache.   Calls to `point_is_within_road` are expensive.